grails-core/src/main/groovy/org/grails/config/NavigableMapConfig.java

```diff
@@ -203,19 +210,33 @@ public String getProperty(String key, String defaultValue) {
 
     @Override
     public <T> T getProperty(String key, Class<T> targetType, T defaultValue) {
-        Object originalValue = configMap.get(key);
-        if(originalValue != null) {
-            if(targetType.isInstance(originalValue)) {
-                return (T)originalValue;
+        // First try to find the property using all registered propertySources
+        PropertySources propertySources = ((PropertySourcesConfig) this).propertySources;
+        if (propertySources != null) {
+            for (PropertySource<?> propertySource : propertySources) {
+                Object originalValue = propertySource.getProperty(key);
+                if (originalValue != null) {
+                    return convertValueIfNecessary(originalValue, targetType, defaultValue);
+                }
             }
-            else {
-                if(!(originalValue instanceof NavigableMap)) {
+        }
```

Comment:
rather than iterating all the property sources which will slow down each property resolution, why not such convert the key (ie `key.toLowerCase(Locale.English).replace('_', '.')`. Wouldn't that work the same?

Reply:
It's not only that logic. Take a look at what Spring is doing: https://github.com/spring-projects/spring-framework/blob/4.3.x/spring-core/src/main/java/org/springframework/core/env/SystemEnvironmentPropertySource.java#L104-L142

And I wanted to avoid copying-pasting all that code but if you think it's better, I can do it.

Reply:
Sorry, now I remember the real reason why I needed to iterate all the property sources: The reason is about **priority**. 

The value for the property in `configMap` is the one defined on  `application.yml` and it's not overridden with the enviroment variables. So, if we search only in `configMap` (no matter if we replace `.`, `_`, `-`) the value returned will always be the one defined in `application.yml` and no the value defined as environment variable when running the application.

Reply:
then you can do a check first for `System.getenv(..)` which will give you whether the environment variable is present.

Reply:
Change done in https://github.com/grails/grails-core/pull/10937/commits/65f7bc6b82103841106a52bb9d35947d2d19662d